File: sdk/python/openwa/resources/catalog.py
"""Catalog resource — WhatsApp Business catalog, products, and product sends.

Backed by ``src/modules/catalog/catalog.controller.ts``
(``@Controller('sessions/:sessionId')``). NOTE: the catalog controller is
mounted under the session root, so catalog reads are ``/catalog...`` while
product sends share the messages namespace
(``/messages/send-product``).
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from .._http import quote_segment
from ..types import (
    CatalogInfo,
    CatalogProduct,
    CatalogProductsQuery,
    PaginatedProducts,
    ProductMessageResponse,
    SendProductRequest,
)

if TYPE_CHECKING:
    from .._http import HttpExecutor


class CatalogResource:
    def __init__(self, http: "HttpExecutor") -> None:
        self._http = http

    def info(self, session_id: str) -> CatalogInfo:
        return self._http.request("GET", f"/api/sessions/{quote_segment(session_id)}/catalog")

    def products(
        self, session_id: str, query: CatalogProductsQuery | None = None
    ) -> PaginatedProducts:
        return self._http.request(
            "GET", f"/api/sessions/{quote_segment(session_id)}/catalog/products", query=query
        )

    def product(self, session_id: str, product_id: str) -> CatalogProduct:
        return self._http.request(
            "GET", f"/api/sessions/{quote_segment(session_id)}/catalog/products/{quote_segment(product_id)}"
        )

    def send_product(self, session_id: str, body: SendProductRequest) -> ProductMessageResponse:
        """Send a product message. Requires an OPERATOR-level key. Shares the messages path."""
        return self._http.request(
            "POST", f"/api/sessions/{quote_segment(session_id)}/messages/send-product", body=body
        )

File: sdk/go/catalog.go
package openwa

import "context"

// CatalogService reads the WhatsApp Business catalog and sends product
// messages. Catalog reads live under /catalog; product sends share the
// messages namespace (/messages/send-product).
// Backed by src/modules/catalog/catalog.controller.ts.
type CatalogService struct{ client *Client }

func (s *CatalogService) sessionBase(sessionID string) string {
	return "/api/sessions/" + pathEscape(sessionID)
}

// Info returns catalog metadata.
func (s *CatalogService) Info(ctx context.Context, sessionID string) (*CatalogInfo, error) {
	var out CatalogInfo
	err := s.client.do(ctx, "GET", s.sessionBase(sessionID)+"/catalog", nil, nil, &out)
	if err != nil {
		return nil, err
	}
	return &out, nil
}

// Products returns paginated catalog products.
func (s *CatalogService) Products(ctx context.Context, sessionID string, query *CatalogProductsQuery) (*PaginatedProducts, error) {
	var out PaginatedProducts
	err := s.client.do(ctx, "GET", s.sessionBase(sessionID)+"/catalog/products", valuesOf(query), nil, &out)
	if err != nil {
		return nil, err
	}
	return &out, nil
}

// Product returns a single catalog product.
func (s *CatalogService) Product(ctx context.Context, sessionID, productID string) (*CatalogProduct, error) {
	var out CatalogProduct
	err := s.client.do(ctx, "GET", s.sessionBase(sessionID)+"/catalog/products/"+pathEscape(productID), nil, nil, &out)
	if err != nil {
		return nil, err
	}
	return &out, nil
}

// SendProduct sends a product message. Requires an OPERATOR-level key.
func (s *CatalogService) SendProduct(ctx context.Context, sessionID string, body SendProductRequest) (*ProductMessageResponse, error) {
	var out ProductMessageResponse
	err := s.client.do(ctx, "POST", s.sessionBase(sessionID)+"/messages/send-product", nil, body, &out)
	if err != nil {
		return nil, err
	}
	return &out, nil
}
